command/agent/monitor/monitor.go

```diff
@@ -0,0 +1,76 @@
+package monitor
+
+import (
+	"sync"
+
+	log "github.com/hashicorp/go-hclog"
+)
+
+type Monitor struct {
+	sync.Mutex
+	sink         log.SinkAdapter
+	logger       log.InterceptLogger
+	logCh        chan []byte
+	index        int
+	droppedCount int
+	bufSize      int
+}
+
+func New(buf int, logger log.InterceptLogger, opts *log.LoggerOptions) *Monitor {
+	sw := &Monitor{
+		logger:  logger,
+		logCh:   make(chan []byte, buf),
+		index:   0,
+		bufSize: buf,
+	}
+
+	opts.Output = sw
+	sink := log.NewSinkAdapter(opts)
+	sw.sink = sink
+
+	return sw
+}
+
+func (d *Monitor) Start(stopCh <-chan struct{}) <-chan []byte {
+	d.logger.RegisterSink(d.sink)
+
+	logCh := make(chan []byte, d.bufSize)
+	go func() {
+		for {
+			select {
+			case log := <-d.logCh:
+				logCh <- log
```

Comment:
Need to do a subselect on the stopCh here to prevent blocking forever if the receiver has gone away. Since we currently put a huge buffer on this chan (512) that's a remote concern, but if someone dramatically reduces the size of the chan they may be caught unaware that it increases the likelihood of a goroutine leak or deadlock.